### PR TITLE
feat: add warning metadata to tables and metrics

### PR DESCRIPTION
**Pull request by etr2460**
### SUMMARY
Similar to the need for certifying metrics and tables, often data should have warnings attached. This provides support for arbitrary warning markdown text to be provided for a metric or table

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
Screen Shot 2021-03-17 at 10 33 39 AM
Screen Shot 2021-03-17 at 10 33 31 AM
Screen Shot 2021-03-17 at 10 33 23 AM

Screen Shot 2021-03-17 at 10 35 55 AM
Screen Shot 2021-03-17 at 10 35 48 AM
Screen Shot 2021-03-17 at 9 52 19 AM

### TEST PLAN
Verify the icon is displayed whenever the deprecated_at param is set. Also add unit tests

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [x] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [x] Introduces new feature or API
- [ ] Removes existing feature or API

to: @graceguo-supercat @ktmud @junlincc (could you tag the appropriate engineers from Preset here?)

### CODE REVIEW

File: superset/connectors/sqla/models.py
Comment (etr2460) [inline superset/connectors/sqla/models.py:499]: ran into this while debugging and got really confused. this seems clearer

**Comment by ktmud** [comment]
@robdiciuccio it seems there's something wrong with the committer detection logic in the Ephemeral environment workflow.

**Comment by robdiciuccio** [comment]
@ktmud yep, turns out your Apache org membership needs to be public to work correctly. There's another possible code fix, but this should hopefully unblock for now.

File: superset-frontend/src/datasource/DatasourceEditor.jsx
Comment (villebro) [inline superset-frontend/src/datasource/DatasourceEditor.jsx:945]: I think having a freetext description field would be more convenient for most orgs. Instead of having a dedicated link field, could we have a general description field in which one can paste links if needed?

**Review by villebro — COMMENTED** [review]
Love the idea! One small question/comment

File: superset-frontend/src/datasource/DatasourceEditor.jsx
Comment (etr2460) [reply superset-frontend/src/datasource/DatasourceEditor.jsx:945]: That's a good suggestion, how would you recommend structuring a description field like that? Ideally, we could provide the capability for someone to embed a link and make it clickable/open in a new tab. Would rendering arbitrary markdown be a bad idea?

Comment (villebro) [reply superset-frontend/src/datasource/DatasourceEditor.jsx:945]: I think we should be ok supporting the same markdown allowances we support on the dashboard markdown component: 
image 
image

I might be oversimplifying, but I'd be happy with just a simple freetext field that's then rendered as markdown.

Comment (etr2460) [reply superset-frontend/src/datasource/DatasourceEditor.jsx:945]: I don't think forcing a new tab in markdown is possible unless you use html `a` tags. i can play around a bit with supporting markdown descriptions. it's a bit less structured (which i'm not sure if that's good or bad).

For context, in internal use, the learn more link typically goes to a google doc or something else with lots of detailed, freeform content. So we made the choice in other products to just link out to there instead of making people write more context/details in product

Comment (ktmud) [reply superset-frontend/src/datasource/DatasourceEditor.jsx:945]: I think we can allow both the free-text markdown and the "Learn more" link, if it's not too much work.

**Review by ktmud — COMMENTED** [review]
Nice feature!

Would it make sense to extend this to more than Deprecation? E.g. It could also be used as a general warning for things like delayed data and suggesting an alternative field:

- "This data has been delayed for x hours. We are working on addressing this issue".
- "This field is for advanced use only. Please use "xxx_xxx" for simpler categorization."

File: superset/connectors/sqla/models.py
Comment (bkyryliuk) [inline superset/connectors/sqla/models.py:429]: it may be nice to have deprecated_by field as well to have point of contact

File: superset-frontend/src/datasource/DatasourceEditor.jsx
Comment (villebro) [reply superset-frontend/src/datasource/DatasourceEditor.jsx:945]: I found this, and making the following change seems to send the links to `_blank`: 
image

```patch
diff --git a/superset-frontend/src/dashboard/components/gridComponents/Markdown.jsx b/superset-frontend/src/dashboard/components/gridComponents/Markdown.jsx
index b071a9982..5b30cd400 100644
--- a/superset-frontend/src/dashboard/components/gridComponents/Markdown.jsx
+++ b/superset-frontend/src/dashboard/components/gridComponents/Markdown.jsx
@@ -286,6 +286,7 @@ class Markdown extends React.PureComponent {
             isValidNode: node => node.type !== 'script',
           }),
         ]}
+        linkTarget="_blank"
       />
     );
   }
```

**Comment by etr2460** [comment]
A bunch of good suggestions here, thanks for the comments!

I think if we're going to provide the flexibility of free text, then we should let the user add whatever content they want there (links, POCs, descriptions, replacements, etc.) and not worry about structuring it more.

However, making this a generic warning might be even better. If we want to take the flexible route, then we could get rid of the deprecated_at field too and just rely on everything to be provided in markdown. Then it's just one simple field that anyone could use for whatever purpose. Super extensible for other users, and would serve the deprecation use case.

I'm going to simplify and move forward with a generic `warning_markdown` field that can be added to a metric or dataset. Will tag y'all when that's ready

**Comment by etr2460** [comment]
Here's the new user experience with markdown: 
Screen Shot 2021-03-15 at 9 21 33 PM
Screen Shot 2021-03-15 at 9 23 39 PM
Screen Shot 2021-03-15 at 9 25 25 PM

**Review by ktmud — COMMENTED** [review]
Thanks for the updates! Code LGTM, tested it locally and found a couple of UX issues:

1.  Is the extra bottom padding intentional?
    Screen Shot 2021-03-16 at 10 31 35 AM

2. Just remembered we already have [`warning_text`](https://github.com/apache/superset/blob/1b95ed726720c17ff74f368357c5ab9c263d59f1/superset/connectors/base/models.py#L591) for metrics, can we maybe migrate it to the same `extra.warning_markdown` field?

    Screen Shot 2021-03-16 at 10 32 39 AM

3. The icon for metric warning markdown does not show up in any of the Dataset panel, Select options, or selected options. Maybe we should add all of them in a separate PR and only do datasource warning in this PR.

    It does show up, however, in the Datasource editor modal.
    Screen Shot 2021-03-16 at 10 34 54 AM

    BTW, there are like 3 different warning icons in this screen. Should we consolidate and make them more consistent?

----

Not related to this PR, but I think we also need some UI adjustment in the datasource editor modal:

image

1. Move "extra" to "Advanced"
2. Make the AceEditor for `extra` taller and resizable
3. Move the two column layout to be two rows with Basic options on top and Advanced in the bottom.
4. Add two column layout __within__ the Basic and Advanced container

cc @junlincc @mihir174

**Comment by etr2460** [comment]
Here are the follow ups i'll make here:
- Use the yellow caution triangle icon for this to be consistent with other warning icons
- Update the metrics to pull either warning_markdown or warning_text (to maintain backwards compatibility)
- Add in a superset-ui PR and dep here which adds the same caution icon for all metric locations

Here's what i'll do in future PRs:
- DB migration to drop the warning_text from the metric model/table
- Update superset-ui to not pull the warning_text anymore

**Comment by junlincc** [comment]
@ktmud 

- Move "extra" to "Advanced" 
Definitely! 
- Make the AceEditor for extra taller and resizable 
Make sense 
-Move the two column layout to be two rows with Basic options on top and Advanced in the bottom.
-Add two column layout within the Basic and Advanced container
Good idea! We probably need more exploration than simply changing the layout. 

@etr2460 Thank you, Erik, for opening this PR and adjusting the UI element for consistency. Really appreciate it!

**Comment by etr2460** [comment]
> Term Data entity is not being used anywhere, should we change it?

@junlincc That's part of the user provided string, and isn't actually in the code anywhere. FWIW, we use `data entity` internally to refer to any of our data objects; it's basically a catch all term to include tables, metrics, columns, datasources, datasets, etc. But since we have no need for that in superset yet, it won't be added

**Comment by etr2460** [comment]
I've made the changes mentioned above, but it won't build until https://github.com/apache-superset/superset-ui/pull/1011 is merged and we have a new superset-ui version

**Comment by etr2460** [comment]
@ktmud i think this is ready to go now

**Review by ktmud — APPROVED** [review]
LGTM!  Could you also add a screenshot of how the icon looks like in the "SAVED" tab of metrics select popover (if it does appear there)?

**Comment by etr2460** [comment]
>Could you also add a screenshot of how the icon looks like in the "SAVED" tab of metrics select popover (if it does appear there)

@ktmud 

image

**Pull request merged**
